crates/fuel-gas-price-algorithm/src/v1.rs

```diff
@@ -372,19 +368,18 @@ impl AlgorithmUpdaterV1 {
             .saturating_div(100)
     }
 
-    fn da_block_update(
-        &mut self,
-        height: u32,
-        block_bytes: u64,
-        block_cost: u64,
-    ) -> Result<(), Error> {
+    fn da_block_update(&mut self, height: u32, block_cost: u64) -> Result<(), Error> {
         let expected = self.da_recorded_block_height.saturating_add(1);
         if height != expected {
             Err(Error::SkippedDABlock {
                 expected: self.da_recorded_block_height.saturating_add(1),
                 got: height,
             })
         } else {
+            let block_bytes = self
```

Comment:
Maybe we want to remove all blocks until `height`?

Reply:
> Maybe we want to remove all blocks until height?

That might be more performant (and kinda what we were doing before). We'd want to do that before we called `da_block_update` and pair each `block_bytes` with the corresponding `RecordedBlock`.

I honestly don't know the performance of `remove` for `HashMap`. `get` is O(1), but obviously doesn't `mut` so `remove` probably does re-scaling and other heap garbage. The most performant would be a `VecDeque` and `split_off` probably? The order is now an issue then. I was going to say we don't need to include the `height` (just the `bytes`) if we trust the order, but just in case we probably should and throw an error if it doesn't match the `recorded_block` it is paired with.

Reply:
> Anyway, leaving this comment for consideration as we might think about protecting unrecorded_blocks from growing indefinitely in case of some unexpected flow.

Yeah. It's directly relevant to what we're talking about. We definitely make assumption about the order.

Talking to the rollup team guys, it sounds like they might not want to guarantee order in the long run, if that's the case then the `HashMap` approach might be the best. We could even get rid of `SkippedL2Block` I think actually.

Reply:
I meant: Block committer submits blocks in the bundles. So bundle can have several blocks inside. When you receive the notification from the block commuter about DA submission, you can have `unrecorded_blocks = vec![block_height-5, block_height-4, ... block_height]`. Then instead of removing only one entry, you need to remove all entries up to `block_height`.

If that is something that we want, then `BTreeMap` is better=)

Reply:
We are removing all in the bundle. We are just iterating over the recorded blocks:
```rs
    pub fn update_da_record_data(
        &mut self,
        blocks: &[RecordedBlock],
    ) -> Result<(), Error> {
        for block in blocks {
            self.da_block_update(block.height, block.block_cost)?;
        }
        self.recalculate_projected_cost();
        self.normalize_rewards_and_costs();
        Ok(())
    }
```
What I was saying is we could get them all first in some efficient way:
```rs
    pub fn update_da_record_data(
        &mut self,
        blocks: &[RecordedBlock],
    ) -> Result<(), Error> {
        let bytes = self.get_bytes_for_recorded_blocks(&blocks)?;
        for (block, bytes) in blocks.iter().zip(bytes) {
            self.da_block_update(block.height, block.block_cost, bytes)?;
        }
        self.recalculate_projected_cost();
        self.normalize_rewards_and_costs();
        Ok(())
    }
``` 
or something similar without `zip`.

Yeah maybe `BTreeMap` makes the most sense for the ordered case.

Reply:
`RecordedBlock` data is linked to each other. It just has the latest block height 

cc @rymnc Since he wrote down the interface that we agreed with Rollup team. Maybe I remember it incorrectly)

Reply:
Yes. I was part of those conversations and we've discussed what is required for the algo.

Reply:
This is what I have -
```rust
struct Bundle {
	sequence_number: u64,
	blocks_range: Range<BlockHeight>,
	// The DA block height of the last transaciton in the bundle.
	da_block_height: DaBlockHeight
	// Total cost of all bundles for the whole history.
	total_cost: u256,
	// Total size of all bundles for the whole history.
	total_size: u256,
}

trait CommitterAPI {
	fn get_n_last_bundle(&self, number: u64) -> Result<Bundle>;

	// Range is based on `sequence_number`
	fn get_bundles_by_range(&self, range: Range<u64>) -> Result<Vec<Bundle>>;
}
```

Reply:
Right. Okay. So in that case we can just sum all the block bytes for that range and do a single cost calculation. Sounds like maybe that's what you were suggesting, @xgreenx . I was still under the impression that more processing would happen on the service side, but this actually works fine.